keps/sig-network/20190613-kube-proxy-component-config.md

```diff
@@ -0,0 +1,204 @@
+- There are options that are shared between some (but not all) modes. Specific features of the underlying implementation are common and this happens only within the boundaries of the platform (iptables and ipvs modes for example).
+- Although legacy Linux and Windows user mode proxies are separate code bases, they have a common set of options.
+
+With that in mind, the following measures are proposed:
+- Mode specific structs are consolidated to not use fields from other mode specific structs.
+- Introduce a single combined legacy user mode proxy struct for both Linux and Windows backends.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ...
+modeB: ...
+modeC: ...
+```
+
+### Group platform specific fields together
+
+Platform specific fields need to be grouped together by platform name. This can be done by introducing structs to house them - one per platform (one for Linux, one for Windows, etc.). Then these structs will have an instance in the KubeProxyConfiguration.
+That can allow for users to specify settings for more than one platform per a YAML document, thus allowing for deduplication of platform independent shared fields. This can reduce the possibility of an error for users.
+So far, research shows, that platform dependent settings are also mode dependent too. Hence each platform struct is going to contain a set of pointers to mode structs - those, that are specific to this platform and platform independent modes.
+Selecting a mode of operation is done on a per-platform basis. To select a mode, only one of the mode sub-structures has to be filled in. If no sub-struct is filled in, a default mode of operation is picked along with its default settings.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+platformX: # Only one sub-key can be provided to select a specific mode of operation
+  modeA: ... # modeA is the selected mode of operation for platformX
+platformY:
+  modeB: ... # modeB is the selected mode of operation for platformY
+```
+
+### Split instance specific and shared settings
+
+As a last step, instance specific and shared settings should be split into a couple of different YAML documents. This will allow for such settings to be specified in one or two YAML files, depending on user preferences.
+Preliminary reviews of the impacted settings show, that only settings, that are not platform or mode specific are instance specific. Therefore, the existing platform-mode specific structures are to be preserved in the shared settings, while instance settings are going to be housed in a small flat object without platform or mode distinction.
+Examples of instance specific settings are `HostnameOverride` and the `*BindAddress` fields.
+
+#### Example
+
+```yaml
+apiVersion: kubeproxy.config.k8s.io
+kind: InstanceConfiguration
+hostnameOverride: "LeHost"
+bindAddress: 1.2.3.4
+---
+apiVersion: kubeproxy.config.k8s.io
+kind: SharedConfiguration
```

Comment:
just a Q, have we established naming conventions for API types (possibly should be done in WG component-base).
e.g. kubeadm has InitConfiguration (no component name prefix), yet kubelet has KubeletConfiguration (component name prefixes the config). kubeproxy v1alpha1 has KubeProxyConfiguration.

Reply:
I think, that as kinds should not have a component prefix. It's already given in the API group.
The only exception to this should be the cases, where a component config API group has a single kind in it. In that case it may be more practical to name the kind `KubeProxyConfiguration` as opposed to just `Configuration`.
However, this is just a personal preference of mine and I am almost certain, that it isn't codified anywhere.

Reply:
The prefix is there for historical reasons - all componentconfig used to be in the same API group. 

I did use that as the convention in the Versioned Component Configuration Files doc: https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit# because it's easy to read and was the convention at that point.

But it doesn't really matter. The package and/or API group information is typically close at hand.

`Config` and `InstanceConfig` could work fine as the convention IMO. 

Fingers crossed I'm not contradicting something I said a long time ago :wink: